File: web_fetch.py
# -*- coding: utf-8 -*-
# ============================================================
#   web_fetch.py - 通用网页抓取（仅依赖 requests，无第三方解析库）
#
#   供 财报助手 / 联网搜索 等插件复用：
#     fetch_html(url)              抓 HTML 源码
#     fetch_text(url)              抓网页正文纯文本
#     clean_text(html)             HTML → 正文文本（去脚本/导航噪音）
#     page_title(html)             提取 <title>
#     find_article_links(...)      列表页 → 文章链接（标题+URL）
#     find_pagination(...)         列表页 → 下一页链接
#     same_domain(u1, u2)          是否同域名
#     is_policy_url(url)           是否像政策文件链接
# ============================================================

import html as _html
import html.parser
import re
from urllib.parse import urljoin, urlparse

import requests

HEADERS = {
    "User-Agent": ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                   "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"),
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
}

NL = chr(10)

_NAV_NOISE = ("首页", "无障碍", "网站地图", "English", "联系我们",
              "版权", "ICP备", "政府网站", "客户端", "微博", "微信", "手机版",
              "政务新媒体", "网站标识码", "分享到", "打印本页", "关闭窗口")

POLICY_KW = ("通知", "办法", "规定", "意见", "公告", "细则", "条例", "政策",
             "批复", "答复", "解读", "zhengce", "zcfg", "content", "xxgk", "gongbao")

_ARTICLE_HREF_KW = ("content", "article", "zhengce", "htm", "shtml", "xxgk",
                    "zwgk", "zcfg", "t202", "detail", "news")

_PAGE_ANCHOR_KW = ("下一页", "下页", "next", ">", "»")


class _TextExtractor(html.parser.HTMLParser):
    """极简正文提取：跳过 script/style/noscript，收集可见文本"""
    def __init__(self):
        super().__init__()
        self.parts = []
        self._skip = 0

    def handle_starttag(self, tag, attrs):
        if tag in ("script", "style", "noscript"):
            self._skip += 1

    def handle_endtag(self, tag):
        if tag in ("script", "style", "noscript") and self._skip:
            self._skip -= 1

    def handle_data(self, data):
        if self._skip:
            return
        t = (data or "").strip()
        if t:
            self.parts.append(t)


def clean_text(html_text, max_chars=8000):
    """HTML → 正文：跳过脚本、去重、短行导航噪音过滤、压缩空白、截断"""
    p = _TextExtractor()
    try:
        p.feed(html_text or "")
    except Exception:
        pass
    lines, seen = [], set()
    total = 0
    for raw in p.parts:
        line = re.sub(r"\s+", " ", raw).strip()
        if not line or line in seen:
            continue
        if len(line) < 40 and any(k in line for k in _NAV_NOISE):
            continue
        seen.add(line)
        lines.append(line)
        total += len(line)
        if total >= max_chars:
            break
    return NL.join(lines)


def page_title(html_text):
    """提取 <title>（失败返回空串）"""
    m = re.search(r"<title[^>]*>(.*?)</title>", html_text or "", re.S | re.I)
    if not m:
        return ""
    return _html.unescape(re.sub(r"<[^>]+>", "", m.group(1))).strip()


def fetch_html(url, timeout=10):
    resp = requests.get(url, headers=HEADERS, timeout=timeout)
    resp.encoding = resp.apparent_encoding or "utf-8"
    return resp.text


def fetch_text(url, timeout=10, max_chars=8000):
    return clean_text(fetch_html(url, timeout), max_chars=max_chars)


def _abs(href, base_url):
    href = (href or "").strip()
    if not href or href.startswith(("javascript:", "#", "mailto:")):
        return ""
    return href if href.startswith("http") else urljoin(base_url, href)


def find_article_links(html_text, base_url, limit=10, href_keywords=None):
    """从列表页挑文章链接：标题 8~90 字 + href 含文章关键词"""
    kws = href_keywords or _ARTICLE_HREF_KW
    pattern = re.compile(r'<a[^>]+href="([^"]+)"[^>]*>(.*?)</a>', re.S | re.I)
    results, seen = [], set()
    for href, raw_title in pattern.findall(html_text or ""):
        title = re.sub(r"\s+", " ", _html.unescape(re.sub(r"<[^>]+>", "", raw_title))).strip()
        if not (8 <= len(title) <= 90):
            continue
        if not any(k in href for k in kws):
            continue
        abs_url = _abs(href, base_url)
        if not abs_url or abs_url in seen:
            continue
        seen.add(abs_url)
        results.append((title, abs_url))
        if len(results) >= limit:
            break
    return results


def find_pagination(html_text, base_url, current_url="", limit=5):
    """列表页 → 下一页链接（锚文本含 下一页/next，或 href 形如 index_N）"""
    pattern = re.compile(r'<a[^>]+href="([^"]+)"[^>]*>(.*?)</a>', re.S | re.I)
    out, seen = [], set()
    cur = (current_url or "").strip()
    for href, raw_anchor in pattern.findall(html_text or ""):
        anchor = re.sub(r"\s+", " ", _html.unescape(re.sub(r"<[^>]+>", "", raw_anchor))).strip()
        abs_url = _abs(href, base_url)
        if not abs_url or abs_url == cur or abs_url in seen:
            continue
        is_next = (anchor and any(k in anchor.lower() for k in _PAGE_ANCHOR_KW)) or re.search(r"index[_-]{BS}d+", href, re.I)
        if not is_next:
            continue
        seen.add(abs_url)
        out.append(abs_url)
        if len(out) >= limit:
            break
    return out


def same_domain(u1, u2):
    try:
        return urlparse(u1).netloc == urlparse(u2).netloc
    except Exception:
        return False


def is_policy_url(url):
    return any(k in (url or "") for k in POLICY_KW)

File: DICK-Android/app/src/main/java/com/dick/plugins/WebFetch.kt
package com.dick.plugins

import java.net.HttpURLConnection
import java.net.URL

/** 通用网页抓取（HttpURLConnection + 正则，零第三方依赖）——移植自 Python web_fetch.py */
object WebFetch {
    val UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"
    val NL = 10.toChar().toString()

    private val NAV_NOISE = listOf("首页", "无障碍", "网站地图", "English", "联系我们", "版权",
        "ICP备", "政府网站", "客户端", "微博", "微信", "手机版", "政务新媒体", "网站标识码",
        "分享到", "打印本页", "关闭窗口")

    val POLICY_KW = listOf("通知", "办法", "规定", "意见", "公告", "细则", "条例", "政策",
        "批复", "答复", "解读", "zhengce", "zcfg", "content", "xxgk", "gongbao")

    private val ARTICLE_KW = listOf("content", "article", "zhengce", "htm", "shtml", "xxgk",
        "zwgk", "zcfg", "t202", "detail", "news")

    fun fetchHtml(url: String, timeoutMs: Int = 8000): String {
        return fetchRaw(url, "UTF-8", timeoutMs)
    }

    /** 按指定字符集抓取（如 GBK 的腾讯行情接口） */
    fun fetchRaw(url: String, charsetName: String, timeoutMs: Int = 8000): String {
        val conn = URL(url).openConnection() as HttpURLConnection
        conn.connectTimeout = timeoutMs
        conn.readTimeout = timeoutMs
        conn.setRequestProperty("User-Agent", UA)
        conn.setRequestProperty("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
        try {
            val code = conn.responseCode
            val stream = if (code >= 400) (conn.errorStream ?: conn.inputStream) else conn.inputStream
            val bytes = stream.readBytes()
            return String(bytes, charset(charsetName))
        } finally {
            conn.disconnect()
        }
    }

    fun unescape(s: String): String = s
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", 34.toChar().toString())
        .replace("&#39;", "'")
        .replace("&nbsp;", " ")

    fun stripTags(html: String): String {
        var h = html
        h = Regex("(?is)<script[^>]*>.*?</script>").replace(h, " ")
        h = Regex("(?is)<style[^>]*>.*?</style>").replace(h, " ")
        h = Regex("(?is)<noscript[^>]*>.*?</noscript>").replace(h, " ")
        h = Regex("<[^>]+>").replace(h, " ")
        return unescape(h)
    }

    fun cleanText(html: String, maxChars: Int = 8000): String {
        val lines = LinkedHashSet<String>()
        var total = 0
        for (raw in stripTags(html).lineSequence()) {
            val line = raw.trim()
            if (line.isEmpty() || line in lines) continue
            if (line.length < 40 && NAV_NOISE.any { it in line }) continue
            lines.add(line)
            total += line.length
            if (total >= maxChars) break
        }
        return lines.joinToString(NL)
    }

    fun pageTitle(html: String): String {
        val m = Regex("(?is)<title[^>]*>(.*?)</title>").find(html) ?: return ""
        return unescape(stripTags(m.groupValues[1])).trim()
    }

    fun fetchText(url: String, timeoutMs: Int = 8000, maxChars: Int = 8000): String =
        cleanText(fetchHtml(url, timeoutMs), maxChars)

    private fun absUrl(href: String, baseUrl: String): String {
        val h = href.trim()
        if (h.isEmpty() || h.startsWith("javascript:") || h.startsWith("#") || h.startsWith("mailto:")) return ""
        if (h.startsWith("http")) return h
        val u = URL(URL(baseUrl), h)
        return u.toString()
    }

    private val linkRe = Regex("""<a[^>]+href="([^"]+)"[^>]*>(.*?)</a>""", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))

    fun findArticleLinks(html: String, baseUrl: String, limit: Int = 10, hrefKeywords: List<String>? = null): List<Pair<String, String>> {
        val kws = hrefKeywords ?: ARTICLE_KW
        val out = mutableListOf<Pair<String, String>>()
        val seen = mutableSetOf<String>()
        for (m in linkRe.findAll(html)) {
            val href = m.groupValues[1]
            val title = stripTags(m.groupValues[2]).trim()
            if (title.length !in 8..90) continue
            if (kws.none { it in href }) continue
            val u = absUrl(href, baseUrl)
            if (u.isEmpty() || u in seen) continue
            seen.add(u)
            out.add(title to u)
            if (out.size >= limit) break
        }
        return out
    }

    fun findPagination(html: String, baseUrl: String, currentUrl: String = "", limit: Int = 5): List<String> {
        val out = mutableListOf<String>()
        val seen = mutableSetOf<String>()
        for (m in linkRe.findAll(html)) {
            val href = m.groupValues[1]
            val anchor = stripTags(m.groupValues[2]).trim().lowercase()
            val u = absUrl(href, baseUrl)
            if (u.isEmpty() || u == currentUrl || u in seen) continue
            val isNext = anchor.isNotEmpty() && (anchor.contains("下一页") || anchor.contains("下页") || anchor.contains("next") || anchor.contains(">"))
            if (!isNext && !Regex("""index[_-]\d+""").containsMatchIn(href)) continue
            seen.add(u)
            out.add(u)
            if (out.size >= limit) break
        }
        return out
    }

    fun sameDomain(u1: String, u2: String): Boolean = try {
        URL(u1).host == URL(u2).host
    } catch (_: Exception) {
        false
    }

    fun isPolicyUrl(url: String): Boolean = POLICY_KW.any { it in url }
}
